aclk/agent_cloud_link.c

```diff
@@ -1443,75 +1482,174 @@ int aclk_update_alarm(RRDHOST *host, ALARM_ENTRY *ae)
 /*
  * Parse the incoming payload and queue a command if valid
  */
-int aclk_handle_cloud_request(char *payload)
+static int aclk_handle_cloud_request(struct aclk_request *cloud_to_agent)
 {
-    struct aclk_request cloud_to_agent = {
-        .type_id = NULL, .msg_id = NULL, .callback_topic = NULL, .payload = NULL, .version = 0
-    };
+    errno = 0;
+    ACLK_SHARED_STATE_LOCK;
+    if (unlikely(aclk_shared_state.agent_state == AGENT_INITIALIZING)) {
+        debug(D_ACLK, "Ignoring \"http\" cloud request; agent not in stable state");
+        ACLK_SHARED_STATE_UNLOCK;
+        return 1;
+    }
+    ACLK_SHARED_STATE_UNLOCK;
 
-    if (aclk_stats_enabled) {
-        ACLK_STATS_LOCK;
-        aclk_metrics_per_sample.cloud_req_recvd++;
-        ACLK_STATS_UNLOCK;
+    if (unlikely(cloud_to_agent->version != aclk_shared_state.version_neg)) {
+        error("Received \"http\" message from Cloud with version %d, but ACLK version %d is used", cloud_to_agent->version, aclk_shared_state.version_neg);
+        return 1;
     }
 
+    if (unlikely(!cloud_to_agent->payload)) {
+        error("payload missing");
+        return 1;
+    }
+    
+    if (unlikely(!cloud_to_agent->callback_topic)) {
+        error("callback_topic missing");
+        return 1;
+    }
+    
+    if (unlikely(!cloud_to_agent->msg_id)) {
+        error("msg_id missing");
+        return 1;
+    }
+
+    if (unlikely(aclk_queue_query(cloud_to_agent->callback_topic, NULL, cloud_to_agent->msg_id, cloud_to_agent->payload, 0, 0, ACLK_CMD_CLOUD)))
+        debug(D_ACLK, "ACLK failed to queue incoming \"http\" message");
+
+    // Note: the payload comes from the callback and it will be automatically freed
+    return 0;
+}
+
+// This handles `version` message from cloud used to negotiate
+// protocol version we will use
+static int aclk_handle_version_response(struct aclk_request *cloud_to_agent)
+{
+    int version = -1;
+    errno = 0;
+
+    if(unlikely(cloud_to_agent->version != ACLK_VERSION_NEG_VERSION)) {
+        error("Unsuported version of \"version\" message from cloud. Expected %d, Got %d", ACLK_VERSION_NEG_VERSION, cloud_to_agent->version);
+        return 1;
+    }
+    if(unlikely(!cloud_to_agent->min_version)) {
+        error("Min version missing or 0");
+        return 1;
+    }
+    if(unlikely(!cloud_to_agent->max_version)) {
+        error("Max version missing or 0");
+        return 1;
+    }
+    if(unlikely(cloud_to_agent->max_version < cloud_to_agent->max_version)) {
```

Comment:
Typo! max should be min

Reply:
thanks this is a very very good catch... hard to spot